src/vmemcache_index.c

```diff
@@ -35,51 +35,130 @@
  */
 
 #include <alloca.h>
+#include <stdlib.h>
 
 #include "vmemcache.h"
 #include "vmemcache_index.h"
 #include "critnib.h"
 
+/* must be a power of 2 */
+#define NSHARDS 256
+
+struct index {
+	struct critnib *bucket[NSHARDS];
+	int sharding;
+};
+
+/*
+ * shard_id -- (internal) hash the key and pick a shard bucket id
+ */
+static int
+shard_id(size_t key_size, const char *key)
+{
+	/* Fowler–Noll–Vo hash */
+	uint64_t h = 0xcbf29ce484222325;
+	for (size_t i = 0; i < key_size; i++)
+		h = (h ^ (unsigned char)*key++) * 0x100000001b3;
+
+	return h & (NSHARDS - 1);
+}
+
+/*
+ * shard -- (internal) pick a shard bucket
+ */
+static struct critnib *
+shard(struct index *index, size_t key_size, const char *key)
+{
+	if (index->sharding)
+		return index->bucket[shard_id(key_size, key)];
+
+	return index->bucket[0];
+}
+
+/*
+ * env_yn10 -- (internal) check an env var for 1/0/y/n, fatal if invalid
+ */
+static int
+env_yesno10(const char *var, int def)
+{
+	const char *q = getenv(var);
+	if (!q)
+		return def;
+
+	if (!strcasecmp(q, "0") || !strcasecmp(q, "n"))
+		return 0;
+	if (!strcasecmp(q, "1") || !strcasecmp(q, "y"))
+		return 1;
+
+	ERR("env var %s needs to be 0 or 1", var);
+	return def;
+}
+
 /*
  * vmcache_index_new -- initialize vmemcache indexing structure
  */
-vmemcache_index_t *
+struct index *
 vmcache_index_new(void)
 {
-	struct critnib *c = critnib_new();
-	if (c)
+	struct index *index = malloc(sizeof(struct index));
+	if (!index)
+		return NULL;
+
+	index->sharding = env_yesno10("VMEMCACHE_SHARDING", 1);
```

Comment:
Just a suggestion: it would be easier to simply have VMEMCACHE_NSHARDS variable, which could be then set to `1` to 'disable' sharding.

BTW, I'm not sure if using ultima-online-lore-derived variable names is a good idea ;) Sharding usually refers to splitting the data across multiple servers (shards, e.g., ultima worlds), whereas what you did is just key partitioning.

Reply:
I never heard about ultima-online — but neither have I heard about this kind of partitioning ever called something other than "sharding".

Reply:
https://www.raphkoster.com/2009/01/08/database-sharding-came-from-uo/
https://en.wikipedia.org/wiki/Shard_(database_architecture)
https://en.wikipedia.org/wiki/Partition_(database)#Partitioning_methods
Your choice, I don't mind either way.